### ISSUE TITLE
第三章 《损失函数》讨论区

### ISSUE THREAD

**Issue by xiaowuhu** [microsoft/ai-edu#414]

**Comment by hellokay** [comment]
0.3-1 均方差损失函数的部分code存在问题。
CalculateCostWB中3D绘图时W,B 应该为网格点坐标矩阵，Loss矩阵格式应该为B行W列。
具体修改如下：
def CalculateCostWB(x,y,n,w,b):
    W = np.linspace(w-2,w+2, 50)
    B = np.linspace(b-2,b+2, 50)
    Loss=np.zeros((len(B),len(W)))
    for i in range(len(B)):
        for j in range(len(W)):
            w = W[j]
            b = B[i]
            a = w*x+b
            loss = CostFunction(x,y,a,n)
            Loss[i,j] = loss
    W, B = np.meshgrid(W, B)
    fig = plt.figure()
    ax = Axes3D(fig)
    ax.plot_surface(W,B,Loss, cmap='rainbow', norm=LogNorm())
    ax.contour(W, B, Loss, zdir = 'z', offset = 0)
    plt.show()

**Comment by xiaowuhu** [comment]
您能把您绘出的图像贴上来吗？我测试得到的是一个平面图。另外，用W当横坐标还是纵坐标，在展示loss的时候区别不大，不知道您的理由是什么？

**Comment by hellokay** [comment]
一、看文章代码时，认为CalculateCostWB和show_3d_surface两个函数描述的是同一个事情，所以它们两个的图像应该是一致的。
二、是否是我理解有所偏差，CalculateCostWB实际描述的是其它的事情？


1. 下面是CalculateCostWB的原始函数生成的图像，它是一个平面。

2. 下面是show_3d_surface生成的图像，它是一个三维图像。
3. 将CalculateCostWB的参数和show_3d_surface的参数保持一致，得到的图像如下。
￼4. 修改CalculateCostWB之后得到的函数图像。




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "microsoft/ai-edu"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年9月12日(星期六) 上午9:41
收件人:&nbsp;"microsoft/ai-edu"<ai-edu@noreply.github.com&gt;;
抄送:&nbsp;"K"<526413367@qq.com&gt;;"Comment"<comment@noreply.github.com&gt;;
主题:&nbsp;Re: [microsoft/ai-edu] 第三章 《损失函数》讨论区 (#414)





 
您能把您绘出的图像贴上来吗？我测试得到的是一个平面图。另外，用W当横坐标还是纵坐标，在展示loss的时候区别不大，不知道您的理由是什么？

**Comment by xiaowuhu** [comment]
是一回事儿。后来我用show_3d_surface代替CalculateCostWB了，在主函数里注释掉对后者的调用，但是没有删掉该函数。

**Comment by hellokay** [comment]
非常感谢你们出的这个教程，让我对神经网络有了更深的理解。




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "microsoft/ai-edu"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年9月14日(星期一) 中午11:00
收件人:&nbsp;"microsoft/ai-edu"<ai-edu@noreply.github.com&gt;;
抄送:&nbsp;"K"<526413367@qq.com&gt;;"Comment"<comment@noreply.github.com&gt;;
主题:&nbsp;Re: [microsoft/ai-edu] 第三章 《损失函数》讨论区 (#414)





 
是一回事儿。后来我用show_3d_surface代替CalculateCostWB了，在主函数里注释掉对后者的调用，但是没有删掉该函数。

**Comment by xiaowuhu** [comment]
u r welcome! hope to give more feedback for the following self-study. 本书《智能之门》快出版了，高等教育出版社，望关注。

**Comment by hellokay** [comment]
非常期待该书出版！




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "microsoft/ai-edu"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年9月14日(星期一) 中午11:09
收件人:&nbsp;"microsoft/ai-edu"<ai-edu@noreply.github.com&gt;;
抄送:&nbsp;"K"<526413367@qq.com&gt;;"Comment"<comment@noreply.github.com&gt;;
主题:&nbsp;Re: [microsoft/ai-edu] 第三章 《损失函数》讨论区 (#414)





 
u r welcome! hope to give more feedback for the following self-study. 本书《智能之门》快出版了，高等教育出版社，望关注。

**Comment by uaicfs** [comment]
您好！书P34页第1行“最朴素的想法就是用求差值”应为“最朴素的想法就是求差值”。
